File: benchmarks/suite/spectral_norm/bench.rs
fn eval_a(i: usize, j: usize) -> f64 {
    1.0 / ((i + j) * (i + j + 1) / 2 + i + 1) as f64
}

fn eval_a_times_u(u: &[f64], n: usize, au: &mut [f64]) {
    for i in 0..n {
        let mut sum = 0.0;
        for j in 0..n {
            sum += eval_a(i, j) * u[j];
        }
        au[i] = sum;
    }
}

fn eval_at_times_u(u: &[f64], n: usize, atu: &mut [f64]) {
    for i in 0..n {
        let mut sum = 0.0;
        for j in 0..n {
            sum += eval_a(j, i) * u[j];
        }
        atu[i] = sum;
    }
}

fn eval_ata_times_u(u: &[f64], n: usize, atau: &mut [f64]) {
    let mut v = vec![0.0; n];
    eval_a_times_u(u, n, &mut v);
    eval_at_times_u(&v, n, atau);
}

fn main() {
    let n = 1000;
    let mut u = vec![1.0_f64; n];
    let mut v = vec![0.0_f64; n];

    for _ in 0..10 {
        eval_ata_times_u(&u, n, &mut v);
        eval_ata_times_u(&v, n, &mut u);
    }

    let mut v_bv = 0.0;
    let mut vv = 0.0;
    for i in 0..n {
        v_bv += u[i] * v[i];
        vv += v[i] * v[i];
    }

    println!("Spectral norm: {:.9}", (v_bv / vv).sqrt());
}

File: benchmarks/suite/spectral_norm/bench.zig
const std = @import("std");

fn evalA(i: usize, j: usize) f64 {
    const ij = i + j;
    return 1.0 / @as(f64, @floatFromInt(ij * (ij + 1) / 2 + i + 1));
}

fn evalATimesU(u: []const f64, n: usize, au: []f64) void {
    for (0..n) |i| {
        var sum: f64 = 0.0;
        for (0..n) |j| {
            sum += evalA(i, j) * u[j];
        }
        au[i] = sum;
    }
}

fn evalAtTimesU(u: []const f64, n: usize, atu: []f64) void {
    for (0..n) |i| {
        var sum: f64 = 0.0;
        for (0..n) |j| {
            sum += evalA(j, i) * u[j];
        }
        atu[i] = sum;
    }
}

fn evalAtATimesU(u: []const f64, n: usize, atau: []f64, tmp: []f64) void {
    evalATimesU(u, n, tmp);
    evalAtTimesU(tmp, n, atau);
}

pub fn main() !void {
    var outbuf: [4096]u8 = undefined;
    var fw = std.fs.File.Writer.init(std.fs.File.stdout(), &outbuf);
    const w = &fw.interface;
    defer w.flush() catch {};

    const allocator = std.heap.page_allocator;

    const n: usize = 1000;

    const u = try allocator.alloc(f64, n);
    defer allocator.free(u);
    const v = try allocator.alloc(f64, n);
    defer allocator.free(v);
    const tmp = try allocator.alloc(f64, n);
    defer allocator.free(tmp);

    for (0..n) |i| {
        u[i] = 1.0;
        v[i] = 0.0;
    }

    for (0..10) |_| {
        evalAtATimesU(u, n, v, tmp);
        evalAtATimesU(v, n, u, tmp);
    }

    var vBv: f64 = 0.0;
    var vv: f64 = 0.0;
    for (0..n) |i| {
        vBv += u[i] * v[i];
        vv += v[i] * v[i];
    }

    try w.print("Spectral norm: {d:.9}\n", .{@sqrt(vBv / vv)});
}
